internal/rbd/rbd_healer.go

```diff
@@ -0,0 +1,184 @@
+	}
+	userKey, ok := secret.Data["userKey"]
+	if !ok {
+		util.ErrorLogMsg("parse userKey failed, %v", ok)
+		return deviceSecret, errors.New("failed to parse userKey")
+	}
+	encryptionPassphrase, ok := secret.Data["encryptionPassphrase"]
+	if !ok {
+		util.ErrorLogMsg("parse encryptionPassphrase failed, %v", ok)
+		return deviceSecret, errors.New("failed to parse encryptionPassphrase")
+	}
+	deviceSecret = map[string]string{
+		"encryptionPassphrase": string(encryptionPassphrase),
+		"userID":               string(userID),
+		"userKey":              string(userKey),
+	}
+
+	return deviceSecret, nil
+}
+
+func callNodeStageVolume(ns *NodeServer, c *k8s.Clientset, pv *v1.PersistentVolume, pluginPath string) error {
+	publishContext := make(map[string]string)
+
+	volID := pv.Spec.PersistentVolumeSource.CSI.VolumeHandle
+	if acquired := ns.VolumeLocks.TryAcquire(volID); !acquired {
+		util.ErrorLogMsg(util.VolumeOperationAlreadyExistsFmt, volID)
+		return status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, volID)
+	}
+	defer ns.VolumeLocks.Release(volID)
+
+	stagingParentPath := pluginPath + "/kubernetes.io/csi/pv/" +
+		pv.Name + "/globalmount"
+
+	util.DefaultLog("sending nodeStageVolume for volID: %s, stagingPath: %s",
+		volID, stagingParentPath)
+
+	deviceSecret, err := getSecret(c,
+		pv.Spec.PersistentVolumeSource.CSI.NodeStageSecretRef.Namespace,
+		pv.Spec.PersistentVolumeSource.CSI.NodeStageSecretRef.Name)
+	if err != nil {
+		util.ErrorLogMsg("getSecret failed for volID: %s, err: %v", volID, err)
+		return err
+	}
+
+	volumeContext := pv.Spec.PersistentVolumeSource.CSI.VolumeAttributes
+	volumeContext["volumeHealerContext"] = "true"
+
+	req := &csi.NodeStageVolumeRequest{
+		VolumeId:          volID,
+		PublishContext:    publishContext,
+		StagingTargetPath: stagingParentPath,
+		VolumeCapability: &csi.VolumeCapability{
+			AccessMode: &csi.VolumeCapability_AccessMode{
+				Mode: accessModeStrToInt(string(pv.Spec.AccessModes[0])),
+			},
+		},
+		Secrets:       deviceSecret,
+		VolumeContext: volumeContext,
+	}
+	if pv.Spec.PersistentVolumeSource.CSI.FSType == "block" {
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Block{
+			Block: &csi.VolumeCapability_BlockVolume{},
+		}
+	} else {
+		req.VolumeCapability.AccessType = &csi.VolumeCapability_Mount{
+			Mount: &csi.VolumeCapability_MountVolume{
+				FsType:     pv.Spec.PersistentVolumeSource.CSI.FSType,
+				MountFlags: pv.Spec.MountOptions,
+			},
+		}
+	}
+
+	_, err = ns.NodeStageVolume(context.TODO(), req)
+	if err != nil {
+		util.ErrorLogMsg("nodeStageVolume request failed, volID: %s, stagingPath: %s, err: %v",
+			volID, stagingParentPath, err)
+		return err
+	}
+
+	return nil
+}
+
+// RunVolumeHealer heal the attached volumes.
+func RunVolumeHealer(ns *NodeServer, conf *util.Config) error {
+	c := util.NewK8sClient()
+	val, err := c.StorageV1().VolumeAttachments().List(context.TODO(), metav1.ListOptions{})
```

Comment:
listoptions should somehow include the node-id, so that only VolumeAttachments for *this node* are returned.

Reply:
Grr! how did I missed this? Maybe my be because I developed this on minikube (single node)?

This is actually a very good catch, thanks @nixpanic

Reply:
Ouch! There seems to be some issue using the field selector for nodeName with volumeattachments,

https://github.com/kubernetes/kubernetes/issues/88387

```
[pkalever@dhcp53-201 ceph-csi]$ kubectl get volumeattachment --field-selector="spec.nodename=minikube"
Error from server (BadRequest): Unable to find "storage.k8s.io/v1, Resource=volumeattachments" that match label selector "", field selector "spec.nodename=minikube": "spec.nodename" is not a known field selector: only "metadata.name", "metadata.namespace"
```

```
E0603 14:02:56.017955 1772482 rbd_healer.go:149] list volumeAttachments failed, err: "spec.nodeName" is not a known field selector: only "metadata.name", "metadata.namespace"
```

It would've been easier with just:
```
val, err := c.StorageV1().VolumeAttachments().List(context.TODO(), metav1.ListOptions{
    FieldSelector: fields.OneTermEqualSelector("spec.nodeName", conf.NodeID).String(),
})                                                                                    
if err != nil {                                                                       
    util.ErrorLogMsg("list volumeAttachments failed, err: %v", err)                   
    return err                                                                        
}                                                                                     
```

Now, we need to filter it in the for loop of List Items[]